dart/collision/CollisionDetector.h

```diff
@@ -43,169 +43,158 @@
+#include "dart/collision/Result.h"
+#include "dart/collision/SmartPointer.h"
 #include "dart/dynamics/SmartPointer.h"
 
 namespace dart {
 namespace collision {
 
-/// Contact information
-struct Contact {
-  // To get byte-aligned Eigen vectors
-  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+class CollisionObject;
 
-  /// Contact point w.r.t. the world frame
-  Eigen::Vector3d point;
+class CollisionDetector : public std::enable_shared_from_this<CollisionDetector>
+{
+public:
 
-  /// Contact normal vector from bodyNode2 to bodyNode1 w.r.t. the world frame
-  Eigen::Vector3d normal;
+  friend class CollisionObject;
+  friend class CollisionGroup;
 
-  /// Contact force acting on bodyNode1 w.r.t. the world frame
-  ///
-  /// The contact force acting on bodyNode2 is -force, which is the opposite
-  /// direction of the force.
-  Eigen::Vector3d force;
+  /// Return collision detection engine type in std::string
+  virtual const std::string& getType() const = 0;
 
-  /// First colliding body node
-  dynamics::WeakBodyNodePtr bodyNode1;
+  /// Create a collision group
+  virtual std::shared_ptr<CollisionGroup> createCollisionGroup() = 0;
 
-  /// Second colliding body node
-  dynamics::WeakBodyNodePtr bodyNode2;
+  /// Create a collision group from any objects that are supported by
+  /// CollisionGroup::registerShapeFramesOf(). Currently, the supporting objects
+  /// are ShapeFrame, std::vector<ShapeFrame>, CollisionGroup, BodyNode, and
+  /// Skeleton.
+  template <typename... Args>
+  std::shared_ptr<CollisionGroup> createCollisionGroup(const Args&... args);
 
-  /// First colliding shape of the first body node
-  dynamics::ShapePtr shape1;
+  /// Perform collision detection for group.
+  virtual bool detect(CollisionGroup* group,
+                      const Option& option, Result& result) = 0;
 
-  /// Second colliding shape of the first body node
-  dynamics::ShapePtr shape2;
+  /// Perform collision detection for group1-group2.
+  virtual bool detect(CollisionGroup* group1, CollisionGroup* group2,
+                      const Option& option, Result& result) = 0;
 
-  /// Penetration depth
-  double penetrationDepth;
+protected:
 
-  // TODO(JS): triID1 will be deprecated when we don't use fcl_mesh
-  /// \brief
-  int triID1;
+  // We define following collision object managers outside of this class for
+  // better code readability.
+  class CollisionObjectManager;
+  class NaiveCollisionObjectManager;
+  class RefCountingCollisionObjectManager;
 
-  // TODO(JS): triID2 will be deprecated when we don't use fcl_mesh
-  /// \brief
-  int triID2;
+  /// Constructor
+  CollisionDetector() = default;
 
-  // TODO(JS): userData is an experimental variable.
-  /// \brief User data.
-  void* userData;
-};
+  /// Claim CollisionObject associated with shapeFrame. New CollisionObject
+  /// will be created if it hasn't created yet for shapeFrame.
+  CollisionObject* claimCollisionObject(const dynamics::ShapeFrame* shapeFrame);
```

Comment:
The `claim`/`unclaim` logic looks *very much* like manual memory management. Could we replace this logic with RAII by using an `std::shared_ptr` to manage `CollisionObject` instances?

Here's a sketch:
- `CollisionDetector` maintains a `std::map<const dynamics::ShapeFrame *, std::weak_ptr<CollisionObject>> mCollisionObjects` of all known `CollisionObject`s internaly.
- `claimCollisionObject` returns `mCollisionObjects[shapeFrame].lock()` if possible.
- If not, it creates a new `CollisionObject`, returns the `shared_ptr`, and stores it in the `mCollisionObjects` map.
- Optionally, we can use a custom deleter on `std::shared_ptr` to clean up internal state in `CollisionDetector`.